File: apps/native/lib/services/inactivityMonitor.ts
import { AppState, AppStateStatus } from 'react-native';
import { useEffect, useRef, useCallback } from 'react';

export interface InactivityMonitorOptions {
  timeout: number; // in milliseconds
  onTimeout: () => void;
  enabled?: boolean;
}

export class InactivityMonitor {
  private timer: NodeJS.Timeout | null = null;
  private lastActivity: Date = new Date();
  private isActive: boolean = true;
  private appStateSubscription: any = null;

  constructor(
    private timeout: number,
    private onTimeout: () => void
  ) {
    this.startMonitoring();
  }

  /**
   * Start monitoring for inactivity
   */
  private startMonitoring(): void {
    // Monitor app state changes
    this.appStateSubscription = AppState.addEventListener(
      'change',
      this.handleAppStateChange.bind(this)
    );

    this.startTimer();
  }

  /**
   * Handle app state changes
   */
  private handleAppStateChange(nextAppState: AppStateStatus): void {
    if (nextAppState === 'active') {
      // App came to foreground
      const inactiveTime = Date.now() - this.lastActivity.getTime();

      if (inactiveTime >= this.timeout) {
        // User was away too long
        this.isActive = false;
        this.onTimeout();
      } else {
        // Resume monitoring
        this.resume();
      }
    } else if (nextAppState === 'background' || nextAppState === 'inactive') {
      // App went to background
      this.lastActivity = new Date();
      this.stopTimer();
    }
  }

  /**
   * Start or restart the inactivity timer
   */
  startTimer(): void {
    this.stopTimer();
    this.lastActivity = new Date();
    this.isActive = true;

    this.timer = setTimeout(() => {
      if (this.isActive) {
        this.isActive = false;
        this.onTimeout();
      }
    }, this.timeout);
  }

  /**
   * Reset the timer on user activity
   */
  resetTimer(): void {
    if (this.isActive) {
      this.startTimer();
    }
  }

  /**
   * Stop the timer
   */
  stopTimer(): void {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    this.isActive = false;
  }

  /**
   * Resume monitoring after being paused
   */
  resume(): void {
    if (!this.isActive) {
      this.isActive = true;
      this.startTimer();
    }
  }

  /**
   * Get time since last activity
   */
  getTimeSinceLastActivity(): number {
    return Date.now() - this.lastActivity.getTime();
  }

  /**
   * Check if currently monitoring
   */
  getIsActive(): boolean {
    return this.isActive;
  }

  /**
   * Clean up resources
   */
  dispose(): void {
    this.stopTimer();
    if (this.appStateSubscription) {
      this.appStateSubscription.remove();
      this.appStateSubscription = null;
    }
  }
}

/**
 * React hook for inactivity monitoring
 */
export function useInactivityMonitor(options: InactivityMonitorOptions) {
  const monitorRef = useRef<InactivityMonitor | null>(null);

  const resetTimer = useCallback(() => {
    if (monitorRef.current) {
      monitorRef.current.resetTimer();
    }
  }, []);

  useEffect(() => {
    if (options.enabled !== false) {
      monitorRef.current = new InactivityMonitor(options.timeout, options.onTimeout);

      return () => {
        if (monitorRef.current) {
          monitorRef.current.dispose();
          monitorRef.current = null;
        }
      };
    }
  }, [options.timeout, options.onTimeout, options.enabled]);

  return { resetTimer };
}

File: apps/mobile/lib/features/auth/services/inactivity_monitor.dart
import 'dart:async';
import 'package:flutter/widgets.dart';

/// Service to monitor user inactivity and trigger timeout actions
class InactivityMonitor extends WidgetsBindingObserver {
  Timer? _timer;
  final Duration timeout;
  final VoidCallback onTimeout;
  DateTime? _lastActivity;
  bool _isActive = true;

  InactivityMonitor({
    required this.timeout,
    required this.onTimeout,
  }) {
    // Start monitoring app lifecycle
    WidgetsBinding.instance.addObserver(this);
    startTimer();
  }

  /// Start or restart the inactivity timer
  void startTimer() {
    _timer?.cancel();
    _lastActivity = DateTime.now();
    _isActive = true;

    _timer = Timer(timeout, () {
      if (_isActive) {
        _isActive = false;
        onTimeout();
      }
    });
  }

  /// Reset the timer on user activity
  void resetTimer() {
    if (_isActive) {
      startTimer();
    }
  }

  /// Stop the timer
  void stopTimer() {
    _timer?.cancel();
    _isActive = false;
  }

  /// Resume monitoring after being paused
  void resume() {
    if (!_isActive) {
      _isActive = true;
      startTimer();
    }
  }

  /// Get time since last activity
  Duration? get timeSinceLastActivity {
    if (_lastActivity == null) return null;
    return DateTime.now().difference(_lastActivity!);
  }

  /// Check if currently monitoring
  bool get isActive => _isActive;

  /// Handle app lifecycle changes
  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    switch (state) {
      case AppLifecycleState.resumed:
        // App is visible and responding to user input
        if (_lastActivity != null) {
          final inactiveTime = DateTime.now().difference(_lastActivity!);
          if (inactiveTime >= timeout) {
            // User was away too long, trigger timeout
            _isActive = false;
            onTimeout();
          } else {
            // Resume monitoring
            resume();
          }
        } else {
          resume();
        }
        break;

      case AppLifecycleState.paused:
      case AppLifecycleState.inactive:
        // App is not visible or responding to user input
        // Record the time when app went to background
        _lastActivity = DateTime.now();
        stopTimer();
        break;

      case AppLifecycleState.detached:
        // App is still hosted but detached from any host views
        stopTimer();
        break;

      case AppLifecycleState.hidden:
        // App is hidden (iOS 14+ and Android 10+)
        _lastActivity = DateTime.now();
        stopTimer();
        break;
    }
  }

  /// Clean up resources
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    _timer?.cancel();
  }
}

/// Widget that wraps content and monitors for user interaction
class InactivityWrapper extends StatefulWidget {
  final Widget child;
  final Duration timeout;
  final VoidCallback onTimeout;
  final bool enabled;

  const InactivityWrapper({
    Key? key,
    required this.child,
    required this.timeout,
    required this.onTimeout,
    this.enabled = true,
  }) : super(key: key);

  @override
  State<InactivityWrapper> createState() => _InactivityWrapperState();
}

class _InactivityWrapperState extends State<InactivityWrapper> {
  late InactivityMonitor _monitor;

  @override
  void initState() {
    super.initState();
    if (widget.enabled) {
      _monitor = InactivityMonitor(
        timeout: widget.timeout,
        onTimeout: widget.onTimeout,
      );
    }
  }

  @override
  void didUpdateWidget(InactivityWrapper oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.enabled != oldWidget.enabled) {
      if (widget.enabled) {
        _monitor = InactivityMonitor(
          timeout: widget.timeout,
          onTimeout: widget.onTimeout,
        );
      } else {
        _monitor.dispose();
      }
    }
  }

  @override
  void dispose() {
    if (widget.enabled) {
      _monitor.dispose();
    }
    super.dispose();
  }

  void _handleUserInteraction() {
    if (widget.enabled) {
      _monitor.resetTimer();
    }
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      behavior: HitTestBehavior.translucent,
      onTap: _handleUserInteraction,
      onPanDown: (_) => _handleUserInteraction(),
      onScaleStart: (_) => _handleUserInteraction(),
      child: MouseRegion(
        onHover: (_) => _handleUserInteraction(),
        child: widget.child,
      ),
    );
  }
}